ZEPXXX-priority-election.md

```diff
@@ -0,0 +1,229 @@
+This situation is also inefficient in terms of resource allocation.
+We should always over-provision nodes to get good performance.
+Therefore, to improve the performance of the system and for an optimal resource usage, it is required to distributed the leaders uniformly among the nodes.
+
+Although the ideal state is to have a strictly uniform distribution of leaders, it is often not possible. A node restart triggers fail-over and re-distribution of the leaders.
+When this node comes back it is highly likely that it cannot become the leader, because it's log is not up-to-date.
+Forcing a re-distribution of leaders at this time is also not ideal, because during the fail-over the system is partly non-available.
+Hence, our current goal is not to get strict uniform distribution. We aim for a best-effort distribution of leaders.
+
+Expected outcome:
+ * After a cluster start or restarts, we aim to get more or less uniformly distributed leaders.
+ * When the current leader dies, the new leaders would be distributed equally in the rest of the nodes in a best-effort way.
+
+Note that it is not always possible to equally distributed the leaders.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We aim to achieve uniform leader distribution by priority based election.
+This means, a node with higher priority has a better chance of becoming the leader.
+
+Each raft node has a `nodePriority` and a `targetPriority`.
+The `nodePriority` is assigned according to its priority to become a leader.
+Higher the `nodePriority`, higher the chance of becoming the leader.
+
+When the `electionTimeout` triggers, instead of immediately sending a poll request, the node first compares the `nodePriority` with the `targetPriority`.
+The node starts the election by sending poll requests only if `nodePriority >= targetPriority`. If the node cannot start the election, it waits for another `electionTimeout`. If there is no new leader elected by that time, the node reduces it's `targetPriority` by `priorityDecay` and repeat the above steps. In this way, the node with the highest priority starts election first and thus has a higher chance of getting elected as the new leader.
+
+Consider the following scenario:  
+NodeA : {nodePriority = 3, targetPriority = 3}  
+NodeB : {nodePriority = 2, targetPriority = 3}  
+NodeC : {nodePriority = 1, targetPriority = 3}  
+
+NodeA is the current leader and it dies. `priorityDecay = 1`.
+
+![priority-election](ZEP006/priority_election.jpg)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The priority election is implemented by the follower.
+Previously, on a heartbeat timeout the follower immediately starts an election by sending a poll request.
+The heartbeat timeout was randomized, in order to prevent multiple nodes starting the election at the same time.
+With the priority election we do not use a randomized heartbeat timeout.
```

Comment:
But using randomized heartbeats would allow us to avoid an election loop, no? What are the downsides of using a randomized heartbeat timeout (within certain bounds)?

Reply:
When randomized timeouts are used it will override the priority, right? We want to enforce a strict order in which the elections are triggered.

Reply:
Will it? We already will have to deal with skewed time - even if we say the time out is 2.5 seconds, not every node will trigger it at the same time. So we already have to deal with some randomization. Say we have a node A with PP 1, and another B with PP 2. B thinks TP is 2, A thinks TP is 1. I guess I sort of expected B to reject a poll request from A for now since it's PP Is lower than what be thinks the TP is, but A would accept one from B.

To some extent this is optimization and can be done later though, as this is probably an edge case.

Reply:
I'm not sure if what you describe is a problematic case. First of all, the poll response is not based on the priority. If B receives a poll or vote requests from A before B vote for itself, it will accept it and A becomes the leader. Election loop will happen when A vote for itself and B vote of itself, and there is no one else to break the tie. If this happens, the election timer is reset and the priority will be reset and it starts from the beginning. So I guess the chances of an election loop is less. (I might be wrong here). 

One idea that I had before was to switch to randomized election if no leader was elected with in a specified time, or if a poll or vote requests fails. But then I thought it was unnecessary as I thought the chances of running into election loop is less with the priority election.